### PR TITLE
[DECOUPLED-MODE] Adding necessary files

### PR DESCRIPTION
**Pull request by gulsumgudukbay**
# Description

This is the first of ~3-4 PRs that is intended for decoupling MaxText from GCE dependencies. 
This PR adds several files including minimal datasets (that are used for unit tests), four scripts to generate these datasets, requirements files for decoupled mode and a conftest file. 
It also edits README for adding decoupled mode information

# Tests
The requirements files were tested to create a decoupled virtual environment. The data generation scripts are also thoroughly tested and the datasets are ingested in unit tests in decoupled mode successfully. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,6 +32,82 @@ Check out our [Read The Docs site](https://maxtext.readthedocs.io/en/latest/) or
 
 See our installation guide to [install MaxText with pip](https://maxtext.readthedocs.io/en/latest/guides/install_maxtext.html).
 
+### Decoupled Mode (No Google Cloud Dependencies)
+
+Set `DECOUPLE_GCLOUD=TRUE` to run MaxText tests and local development without any Google Cloud SDK, `gs://` buckets, JetStream, Tunix, or Vertex AI integrations.
```
Comment (shralex) [inline README.md]: Tunix and Jetstream are OSS libraries rather than a Google Cloud dependencies. Can we exclude them from this list ? I imagine people might want to run Tunix with decoupled mode. I don't have strong feeling about Jetstream, if there's a good reason to keep it on the list.

File: dependencies/requirements/requirements_decoupled_jax_0_7.1.txt
Comment (shralex) [inline dependencies/requirements/requirements_decoupled_jax_0_7.1.txt:1]: If this file is AMD-specific, do we need to merge it?

File: dependencies/requirements/requirements_decoupled_jax_0_6_0.txt
Comment (shralex) [inline dependencies/requirements/requirements_decoupled_jax_0_6_0.txt:1]: Do we need a separate file, or can this be somehow a diff of another requirements file ?

Comment (shralex) [reply dependencies/requirements/requirements_decoupled_jax_0_6_0.txt:1]: I noticed that multiple new requirements are added like chex, etils evaluate, etc. Can you explain why this is needed ?

File: src/MaxText/gcloud_stub.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
+"""Centralized decoupling helpers for JetStream / Tunix / cloud diagnostics / GCS.
```
Comment (shralex) [inline src/MaxText/gcloud_stub.py]: Please add a licensing block to any py files you're adding

File: dependencies/requirements/requirements_decoupled_jax_0_6_0.txt
Comment (gulsumgudukbay) [reply dependencies/requirements/requirements_decoupled_jax_0_6_0.txt:1]: > I noticed that multiple new requirements are added like chex, etils evaluate, etc. Can you explain why this is needed ?

So initially I started to use requirements_with_jax_stable_stack_0_6_1_pipreqs.txt file and modify the requirements accordingly both for JAX 0.6.0, and then any files or packages that require a package, I added them. Some of them are for minimal dataset generations, and some of them are dependencies of other packages, so I pinned them as well. Evaluate exists in this file and also is imported by several different scripts (in src/MaxText/inference_mlperf directory).   

etils is being used by array_record (minimal dataset generation), tensorflow-datasets, orbax-checkpoint and grain packages. 
chex is being used by transformer_engine

Comment (gulsumgudukbay) [reply dependencies/requirements/requirements_decoupled_jax_0_6_0.txt:1]: > Do we need a separate file, or can this be somehow a diff of another requirements file ?

I will remove this one and keep the 0.7.1 one. Since I removed jax dependencies and this was for 0.6.0 JAX, it has outdated and strict pinning so I will remove this.

File: dependencies/requirements/requirements_decoupled_jax_0_7.1.txt
Comment (gulsumgudukbay) [reply dependencies/requirements/requirements_decoupled_jax_0_7.1.txt:1]: It is not, sorry I forgot to rename it. it is generic currently.

File: README.md (hunk 1 above)
Comment (gulsumgudukbay) [reply README.md]: Since Jetstream is currently only for TPUs, added decoupling support for that as well (that means a user can optionally install it and the installed lib will be used if it is available, if not, stub will be used (in decoupled mode)
For tunix: one important dependency that is downloaded when it is in the requirements file is gcsfs. Since that is purely google cloud storage dependency, I avoided it in the requirements file. 

I can edit the README file saying that these are optional - a user can install Jetstream or Junix and if they exist in the environment, they will be used, otherwise gcloud_stub will use empty stubs. 

In fact, the optional behavior is same throughout all dependencies. The gcloud_stub checks if the google cloud dependent dependency exist in the environment. If not, it proceeds with the stub.

File: src/MaxText/gcloud_stub.py (hunk 2 above)
Comment (gulsumgudukbay) [reply src/MaxText/gcloud_stub.py]: I added the block

File: README.md [hunk 3]
```diff
@@ -32,6 +32,79 @@ Check out our [Read The Docs site](https://maxtext.readthedocs.io/en/latest/) or
 
 See our installation guide to [install MaxText with pip](https://maxtext.readthedocs.io/en/latest/guides/install_maxtext.html).
 
+### Decoupled Mode (No Google Cloud Dependencies)
```
Comment (shralex) [inline README.md]: Can we move decoupled mode into a separate doc in docs/guides/run_maxtext folder, and just have something short in the readme pointing to the doc ?

File: local_datasets/c4_en_dataset_minimal/c4/en/3.0.1/__local_c4_builder-train.tfrecord-00000-of-00008
Comment (shralex) [inline local_datasets/c4_en_dataset_minimal/c4/en/3.0.1/__local_c4_builder-train.tfrecord-00000-of-00008:1]: do we need both 3.1.0 and 3.0.1 ?

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  deselected = []
+
+  skip_no_tpu = None
+  if not _HAS_TPU:
+    skip_no_tpu = pytest.mark.skip(reason="Skipped: requires TPU hardware, none detected")
+
+  for item in items:
+    cur_test_markers = {m.name for m in item.iter_markers()} # Iterate thru the markers of every test
+    # Hardware skip retains skip semantics
+    if skip_no_tpu and "tpu_only" in cur_test_markers:
```
Comment (shralex) [inline tests/conftest.py:69]: should we do the same if no gpu ?

File: pytest.ini [hunk 5]
```diff
@@ -16,7 +16,11 @@ markers =
     tpu_only: marks tests to be run on TPUs only
     gpu_only: marks tests to be run on GPUs only
     cpu_only: marks tests to be run on CPUs only
+    decoupled: tests that validate offline / DECOUPLE_GCLOUD=TRUE mode
```
Comment (shralex) [inline pytest.ini]: can you comment that this marker is not to be used manually -- it is applied automatically when the test is not marled as external_*  (if my understanding is correct)

File: local_datasets/c4_en_dataset_minimal/c4/en/3.0.1/__local_c4_builder-train.tfrecord-00000-of-00008
Comment (gulsumgudukbay) [reply local_datasets/c4_en_dataset_minimal/c4/en/3.0.1/__local_c4_builder-train.tfrecord-00000-of-00008:1]: Yes, so some tests directly used 3.0.1 array record format. Some tests required tfrecord formats so I generated them using 3.0.1, and the conversion gave error because the 3.0.1 is not in the mirror anymore. So when I copied 3.0.1 as 3.1.0 (since it was looking for 3.1.0 version), the conversion succeeded.

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      item.add_marker(pytest.mark.decoupled)
+
+
+def pytest_configure(config):
+  for m in [
+      "gpu_only: tests that require GPU hardware",
+      "tpu_only: tests that require TPU hardware",
+      "external_serving: JetStream / serving / decode server components",
+      "external_training: goodput integrations",
+      "decoupled: marked on tests that are not skipped due to GCE deps, when DECOUPLE_GCLOUD=TRUE",
```
Comment (shralex) [inline tests/conftest.py]: GCE -> GCP

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+except Exception:  # pragma: no cover  pylint: disable=broad-exception-caught
+  _HAS_TPU = False
+
+try:
+  _HAS_GPU = any(d.platform == "gpu" for d in jax.devices())
+except Exception:  # pragma: no cover  pylint: disable=broad-exception-caught
+  _HAS_GPU = False
+
+
+GCE_MARKERS = {"external_serving", "external_training"}
```
Comment (shralex) [inline tests/conftest.py]: GCE -> GCP

**Comment by gulsumgudukbay** [comment]
> Thank you for this great contribution!

It was my pleasure!
